### PR TITLE
nix-channel: implement in c++

### PR DESCRIPTION
**Pull request by shlevy**

### CODE REVIEW

File: src/nix-channel/nix-channel.cc [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+    return handleExceptions(argv[0], [&]() {
+        initNix();
+
+        // Turn on caching in nix-prefetch-url.
+        auto channelCache = settings.nixStateDir + "/channel-cache";
+        createDirs(channelCache);
+        setenv("NIX_DOWNLOAD_CACHE", channelCache.c_str(), 1);
+
+        // Figure out the name of the `.nix-channels' file to use
+        auto home = getenv("HOME");
```
Comment (edolstra) [inline src/nix-channel/nix-channel.cc]: => `getEnv` (which returns a std::string).

File: src/nix-channel/nix-channel.cc [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+{
+    auto channelsFD = AutoCloseFD{open(channelsList.c_str(), O_RDONLY | O_CLOEXEC)};
+    if (!channelsFD) {
+        if (errno == ENOENT)
+            return;
+        throw SysError(format("opening ‘%1%’") % channelsList);
+    }
+
+    auto channelsFile = readFile(channelsFD.get());
+    channelsFD = -1;
```
Comment (edolstra) [inline src/nix-channel/nix-channel.cc]: Why not
````
if (!pathExists(channelsList)) return;
auto channelsFile = readFile(channelsList);
```

**Comment by shlevy** [comment]
@edolstra updated

**Comment by edolstra** [comment]
Looks good to me.

**Pull request merged**